R/utilities-sensitivity-calculation.R

```diff
@@ -2,49 +2,217 @@
-    dplyr::group_by(ParameterPath, PKParameter) %>%
+.simulationResultsToPKDataFrame <- function(simulationResults,
+                                            parameterPath,
+                                            customOutputFunctions = NULL) {
+  # calculate standard pkAnalyses
+  pkDataList <- userPKDataList <-
+    setNames(
+      vector("list", length(simulationResults)),
+      names(simulationResults)
+    )
+  for (simResult in names(simulationResults)) {
+    pkDataFrame <- pkAnalysesToDataFrame(
+      calculatePKAnalyses(
+        simulationResults[[simResult]]
+      )
+    )
+    pkDataList[[simResult]] <- pkDataFrame
+  }
+  pkData <- dplyr::bind_rows(pkDataList, .id = "ParameterFactor")
+
+  # calculate user-defined pkAnalyses
+  if (!is.null(customOutputFunctions)) {
+    for (simResult in names(simulationResults)) {
+      userPKDataFrame <- .calculateCustomPK(
+        simulationResults[[simResult]],
+        customOutputFunctions
+      )
+      userPKDataList[[simResult]] <- userPKDataFrame
+    }
+    userPKData <- dplyr::bind_rows(userPKDataList, .id = "ParameterFactor")
+  } else {
+    userPKData <- data.frame()
+  }
+
+  # combine standard and user-defined PK data
+  pkData <- dplyr::bind_rows(pkData, userPKData)
+
+  # modify and format data
+  pkData <- dplyr::rename(
+    pkData,
+    OutputPath       = QuantityPath,
+    PKParameter      = Parameter,
+    PKParameterValue = Value
+  )
+
+  pkData <- .addParameterColumns(pkData, simulationResults, parameterPath)
+  pkData <- dplyr::group_by(pkData, ParameterPath, PKParameter) %>%
     dplyr::group_modify(.f = ~ .computePercentChange(.)) %>%
-    dplyr::ungroup() %>%
-    dplyr::select(
-      "OutputPath",
-      dplyr::starts_with("Parameter"),
-      dplyr::starts_with("PK"),
-      Unit, PercentChangePK,
-      dplyr::everything(),
-      -c("IndividualId")
-    ) %>%
+    dplyr::ungroup()
+
+  pkData <- dplyr::select(
+    pkData,
+    "OutputPath",
+    dplyr::starts_with("Parameter"),
+    dplyr::starts_with("PK"),
+    Unit, PercentChangePK,
+    dplyr::everything(),
+    -c("IndividualId")
+  ) %>%
     dplyr::arrange(ParameterPath, PKParameter, ParameterFactor)
+
+  return(pkData)
+}
+
+#' Calculate custom PK values
+#'
+#' This function calculates user-defined PK values from simulation results using
+#' custom output functions.
+#'
+#' @param simulationResults `SimulationResults` object containing the simulation
+#' data.
+#' @param customOutputFunctions Named list of custom output functions to calculate
+#' PK values.
+#'
+#' @return A dataframe containing the custom PK values.
+#'
+#' @keywords internal
+#' @noRd
+.calculateCustomPK <- function(simulationResults, customOutputFunctions) {
+  # validate customOutputFunctions
+  validateIsOfType(customOutputFunctions, "list", nullAllowed = TRUE)
+  validateIsOfType(customOutputFunctions, "function", nullAllowed = TRUE)
+  validateHasOnlyNonEmptyStrings(names(customOutputFunctions))
```

Comment:
I would maybe create a new validate... function in {esqlabsR} for checking if lists are named. Because the message returned by this validation will not be clear to the user: 
```
> customOutputFunctions <- list(1, 2, 3)
> ospsuite.utils::validateHasOnlyNonEmptyStrings(names(customOutputFunctions))
Error in ospsuite.utils::validateHasOnlyNonEmptyStrings(names(customOutputFunctions)) : 
  ospsuite.utils::validateHasOnlyNonEmptyStrings: argument 'names(customOutputFunctions)' has empty strings!
```

We could return a more explicit message: `customOutputFunctions is not a named list` for example